.github/workflows/docker-publish.yml

```diff
@@ -0,0 +1,51 @@
+name: Docker Publish
+
+on:
+  workflow_run:
+    workflows : ["docker-build"]
+    branches: ["main"]
+    types:
+      - completed
+  push:
+    branches:
+      - "main"
+    tags:
+      - "v*.*.*"
+
+jobs:
+  docker:
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: false
+      matrix:
+        platform:
+          - linux/amd64
+          - linux/arm64
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          submodules: 'recursive'
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Login to Docker Hub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      - name: Get short sha
+        id: sha
+        run: echo "sha_short=$(git rev-parse --short HEAD)" >> $GITHUB_OUTPUT
+      - name: Build and push
+        uses: docker/build-push-action@v5
+        with:
+          context: .
+          platforms: ${{ matrix.platform }}
+          push: true
+          tags: >
+            adfreiburg/qlever:${{ steps.sha.outputs.sha_short }},
+            adfreiburg/qlever:${{ github.ref_name }},
+            adfreiburg/qlever:latest
```

Comment:
1. Did you check about the `latest` tag, whether it's implicit or not, and what happens if you push multiple instances?

2.  Maybe you have seen, that our commit messages typically end with `(#1093)`  which notes the number of the merged pull request. It would be nice to parse this and have one of the
additional tags be something like `qlever:pr-1093` (with a graceful fallback in case something doesn't work).

Reply:
And is it necessary to hardcode `adfreiburg`?
Maybe we can use the dockerhub username there automatically.

(small hint for testing: setup a repo/branch where the docker build only takes seconds (only a dummy), then debugging these workflow things is much easier.

Reply:
1. The `latest` tag will be overwritten by the new image, when it is pushed. So every time this workflow runs, it will overwrite the old `latest` image with the current one.
2. Unfortunately, Github actions provides very limited string operations. AFAIK the most "involved" operation is a string join. And to my knowledge there is no context variable which contains the PR number only. Another problem with this is, that the action is not supposed to be triggered by a PR, but only on a successful merge on master. Within this context, the action does not have access to the PR anymore.

Instead I would strongly advise that you use [semantic versioning](https://semver.org/). It makes it much easier for users to built software that uses Qlever, because semver clearly communicates what change has which effect.
With the publish action, using semver is pretty easy for you. All you need to do is to push a new tag and then create a release from that tag, while the build and publish runs in the background. Then you can also choose, whether every PR should be a new release or you bundle them together.
[I'm also not the first to propose this.](https://github.com/ad-freiburg/qlever/issues/524#issuecomment-994057593)

Reply:
> (small hint for testing: setup a repo/branch where the docker build only takes seconds (only a dummy), then debugging these workflow things is much easier.

I've been using [Act](https://github.com/nektos/act). It emulates Github actions locally, which offers more control how to run jobs. So far, this met my needs and I didn't need a dummy repo. But thanks for the hint, I'll keep it in mind.

Reply:
1. Thanks for the hint with Act, that can indeed be very useful, If I only had known earlier:)
2. As for the PR number: Of course you cannot get to the PR number, as technically there is no PR. But probably you can get to the commit message at least via the git commandline.
And from the commit message (which has the form as our convention)
`Title of the commit message (#10293)`
You can perform a string transformation to transform this into the tag `qlever:pr10293` (in the rare case where this fails, the tag shouldn't be created).
As for string transformations:
If the github internal scripting engine doesn't provide what you want, you can still start a simple `run` step that performs all your transformations on the commandline (using set/python/whatever you like), and then get it back. So all in all I still think, that this is implementable in a feasible way:)